vpr/test/test_connection_router.cpp

```diff
@@ -0,0 +1,131 @@
+#include "vpr_signal_handler.h"
+#include "globals.h"
+#include "net_delay.h"
+#include "place_and_route.h"
+#include "route_tree_timing.h"
+#include "timing_place_lookup.h"
+
+static constexpr const char kArchFile[] = "../../vtr_flow/arch/timing/k6_frac_N10_mem32K_40nm.xml";
+
+namespace {
+
+// Route from source_node to sink_node, returning either the delay, or infinity if unroutable.
+static float do_one_route(int source_node, int sink_node, const t_router_opts& router_opts, const std::vector<t_segment_inf>& segment_inf) {
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& route_ctx = g_vpr_ctx.routing();
+
+    t_rt_node* rt_root = init_route_tree_to_source_no_net(source_node);
+
+    // Update base costs according to fanout and criticality rules.
+    update_rr_base_costs(1);
+
+    // Bounding box includes the entire grid.
+    t_bb bounding_box;
+    bounding_box.xmin = 0;
+    bounding_box.xmax = device_ctx.grid.width() + 1;
+    bounding_box.ymin = 0;
+    bounding_box.ymax = device_ctx.grid.height() + 1;
+
+    t_conn_cost_params cost_params;
+    cost_params.criticality = router_opts.max_criticality;
+    cost_params.astar_fac = router_opts.astar_fac;
+    cost_params.bend_cost = router_opts.bend_cost;
+
+    route_budgets budgeting_inf;
+
+    RouterStats router_stats;
+    auto router_lookahead = make_router_lookahead(
+        router_opts.lookahead_type,
+        router_opts.write_router_lookahead,
+        router_opts.read_router_lookahead,
+        segment_inf);
+
+    ConnectionRouter<BinaryHeap> router(
+        device_ctx.grid,
+        *router_lookahead,
+        device_ctx.rr_nodes,
+        device_ctx.rr_rc_data,
+        device_ctx.rr_switch_inf,
+        g_vpr_ctx.mutable_routing().rr_node_route_inf);
+
+    // Find the cheapest route if possible.
+    bool found_path;
+    t_heap cheapest;
+    std::tie(found_path, cheapest) = router.timing_driven_route_connection_from_route_tree(rt_root, sink_node, cost_params, bounding_box, router_stats);
+
+    // Default delay is infinity, which indicates that a route was not found.
+    float delay = std::numeric_limits<float>::infinity();
+    if (found_path) {
+        // Check that the route goes to the requested sink.
+        REQUIRE(cheapest.index == sink_node);
+
+        // Get the delay
+        t_rt_node* rt_node_of_sink = update_route_tree(&cheapest, OPEN, nullptr);
+        delay = rt_node_of_sink->Tdel;
+
+        // Clean up
+        free_route_tree(rt_root);
+    }
+
+    // Reset for the next router call.
+    router.reset_path_costs();
+    return delay;
+}
+
+// Test that the router can route nets individually, not considering congestion.
+// This is a minimal timing driven routing test that can be used as documentation,
+// and as a starting point for experimentation.
+TEST_CASE("connection_router", "[vpr]") {
+    // Source and sink VPR would have chosen for wire.eblif
+    // Determined by running VPR with the same parameters.
+    int source_rr_node = 198;
```

Comment:
This is dangerous because a change in the rr graph generator would change these numbers.  Consider your own randomize the rr graph.

A couple examples you could do:
 - Run the pack/place for `wire.eblif` to populate the net terminals
 - Pick random SOURCE / SINK from the graph to route
 - Search the cluster list for IO tiles (e.g. model == ".input" / ".output", etc) and use SOURCE and SINK rr nodes from those tiles

Reply:
I've added a more robust method for picking the source and sink by walking the RR graph.